src/daemon.rs

```diff
@@ -112,6 +112,13 @@ struct NetworkInfo {
     relayfee: f64, // in BTC
 }
 
+#[derive(Serialize, Deserialize, Debug)]
+struct IndexInfo {
+    synced: bool,
+    best_block_height: usize,
+}
```

Comment:
I suggest renaming this to `TxIndexInfo`. (or `TxindexInfo`? the former reads better IMO, the latter seems formally consistent)

Then have a helper enum:

```rust
[derive(Deserialize, Debug)]
#[serde(rename_all = "lowercase")]
enum IndexInfo {
    TxIndex(TxIndexInfo),
}
```

Then try to deserialize `Vec<IndexInfo>` directly from the response:

```rust
fn getindexinfo(&self, indexes: Vec<&str>) -> Result<Vec<IndexInfo>> {
    self.request("getindexinfo", json!(indexes)).map_err(Into::into)
}
```

**Why enum instead of struct?**

This is preparation for the possibility of adding other index infos.
If this is upstreamed to `bitcoincore-rpc` it'd be the best way wit additional `#[non_exhaustive]` attribute (but not sure if its MSRV supports it OTOH maybe it can be bumped). It will force us to add `_ => None` match arm below, but that's fine.

I have an alternative black magic idea that would statically guarantee specific type (for our use case) but it's not exclusive because I think the file argument is optional.

Reply:
I really like the enum I am not familiar with it but it really looks much better indeed fo futur index types :)

Reply:
I have trouble with `getindexinfo` because `self.request` return me a `Value` type.

So I deserialize it by returning `from_value(self.request("getindexinfo", json!(indexes))?).chain_err(|| "invalid index info")` is that ok or do you think there is a better way to do it ?

Reply:
It's OK stopgap solution but long term it should be changed to deserialize into the struct directly. This will be probably cleaner in the `bitcoincore-rpc` crate, so let's wait for it with that for now and only try to solve it later.

Although that method should be changed anyway because whatever is using it may be underperformant. Should be a different PR, I believe.

Reply:
> long term it should be changed to deserialize into the struct directly.

Do you mean it requires to change `self.request` ? Any way as you said, if it makes it to `bitcoincore-rpc` then code will looks better, no doubt.

> Although that method should be changed anyway because whatever is using it may be underperformant. Should be a different PR, I believe.

I am not sure if you are talking about using the `TxIndexState` but I am thinking about how to do it now, it is clear that asking for `txindex` value each time we query a transaction is not good :/

Reply:
> Do you mean it requires to change `self.request` ?

Yes something along the lines: `fn request<T: serde::Deserialize>(...) -> Result<T>`

> I am not sure if you are talking about using the `TxIndexState`

No, I meant changing `request()`. Anyway this is `master` branch and I'm not sure if there will be a release before `p2p` gets finished.

> it is clear that asking for `txindex` value each time we query a transaction is not good

Yep, I'm just unsure how to handle failed connection without a huge code rewrite and forgetting something. One thing that comes to mind is wrapping the connection in a helper struct that will contain that cache and any other cache in the future. The struct is initialized with empty values after connecting.